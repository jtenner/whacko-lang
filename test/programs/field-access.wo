class Foo {
    bar: i32;
    constructor() {
        this.bar = i32(123);
    }
}

export fn _start(): void {
    let baz = new Foo();
    assert(baz.bar == i32(123));
}
