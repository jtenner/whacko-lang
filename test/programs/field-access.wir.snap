program:
  callable:[std/builtins.wo]gc.initialize;functionType:()=>voidType
(extern function)
  callable:[std/builtins.wo]gc.alloc;functionType:(u32Type,usizeType)=>ptrType
(extern function)
  callable:[std/builtins.wo]gc.store_field_ptr;functionType:(i32Type,ptrType,ptrType,ptrType)=>voidType
(extern function)
  callable:[test/programs/field-access.wo]_start;functionType:()=>voidType
    block: entry~513
      517~inst = new: classType:[test/programs/field-access.wo]Foo<>
      518~inst = store: (variable compiler-generated) = (runtime classType:[test/programs/field-access.wo]Foo<> 517~inst)
      519~inst = call: callable:[test/programs/field-access.wo]Foo.constructor;methodType:classType:[test/programs/field-access.wo]Foo<>#()=>voidType with (runtime classType:[test/programs/field-access.wo]Foo<> 517~inst)
      520~inst = store: (variable baz) = (runtime classType:[test/programs/field-access.wo]Foo<> 517~inst)
      521~inst = load: (variable baz)
      522~inst = load: (field classType:[test/programs/field-access.wo]Foo<>#bar of (runtime classType:[test/programs/field-access.wo]Foo<> 521~inst))
      527~inst = (i32 123)
      528~inst = 522~inst == 527~inst
      br to next~531 if (runtime i32Type 528~inst); else to falsy~530
    block: falsy~530
      534~inst = unreachable
    block: next~531
      return (void)

  callable:[test/programs/field-access.wo]Foo.constructor;methodType:classType:[test/programs/field-access.wo]Foo<>#()=>voidType
    block: entry~536
      537~inst = load: (variable const compiler-generated)
      538~inst = load: (field classType:[test/programs/field-access.wo]Foo<>#bar of (runtime classType:[test/programs/field-access.wo]Foo<> 537~inst))
      543~inst = (i32 123)
      544~inst = store: (field classType:[test/programs/field-access.wo]Foo<>#bar of (runtime classType:[test/programs/field-access.wo]Foo<> 537~inst)) = (runtime i32Type 543~inst)
      return (void)

